File: python-algorithm/algo_05_dynamic_plan/dp_5_rob/L213-m.py
#--coding:utf-8--

class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 0:
            return 0
        if len(nums) == 1:
            return nums[0]
        # 不偷第一个房子和不偷最后一个房子
        amount1 = self.exec(nums[:-1])
        amount2 = self.exec(nums[1:])
        return max(amount1, amount2)

    def exec(self, nums):
        print(nums)
        sz = len(nums)
        dp = [0] * (sz + 1)
        dp[1] = nums[0]
        for i in range(2, sz + 1):
            # if else 应该括起来，否则运算优先级会改变掉逻辑
            dp[i] = max(dp[i - 1], (dp[i - 2] if i >= 2 else 0) + nums[i - 1])
        return dp[sz]

File: python-algorithm/algo_05_dynamic_plan/dp_5_rob/L213-m.cpp

#include <iostream>
#include <vector>
using namespace std;
class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        if(n==0) return 0;
        if(n == 1) return nums[0];
        int Max1 = rob1(nums,0,n-1);//不偷窃最后一个房子
        int Max2 = rob1(nums,1,n);//不偷窃第一个房子
        return max(Max1,Max2);
    }

    int rob1(vector<int>& nums,int start,int end){
        /*
        动态转移方程是：
        f(n)=max(nums[n]+f(n-2),f(n-1))
        prevMax:f(k-2)
        currMax:f(k-1)
        x:Ak
        */
        int prevMax = 0;
        int currMax = 0;
        for(int i=start;i<end;i++){
            int temp = currMax;
            currMax = max(nums[i]+prevMax, currMax);
            prevMax = temp;
        }
        return currMax;
    }
};